sql/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java

```diff
@@ -118,31 +124,54 @@
             rowShardResolver,
             indexNameResolver,
             expressions,
+            metaExpressions,
             itemFactory,
+            itemFailureFactory,
+            sourceUriFailureFactory,
+            sourceUriFailureInput,
             autoCreateIndices
         );
         bulkShardCreationLimiter = new BulkShardCreationLimiter<>(tableSettings,
             clusterService.state().nodes().getDataNodes().size());
+        this.resultCollector = resultCollector;
     }
 
-    public CompletableFuture<Long> execute(ShardedRequests<ShardUpsertRequest, ShardUpsertRequest.Item> requests) {
+    public CompletableFuture<UpsertResults> execute(ShardedRequests<ShardUpsertRequest, ShardUpsertRequest.Item> requests) {
+        final UpsertResults upsertResults = resultCollector.supplier().get();
+        collectFailingItems(requests, upsertResults);
         if (requests.itemsByMissingIndex.isEmpty()) {
-            return execRequests(requests.itemsByShard);
+            return execRequests(requests.itemsByShard, upsertResults);
```

Comment:
Why do we pass the upsertResults here? Couldn't we create them inside `execRequests`? We return them anyways through the `CompletableFuture`.

Reply:
Ah ok, it collects the failed URIs before.